File: CLI Engine/internal/extractors/usage.go
package extractors

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/monitor/armmonitor"
	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources"
)

// usageWorkerCount bounds how many Azure Monitor calls run concurrently.
// Azure Monitor's default throttling limit is generous enough that this
// stays well clear of 429s while being far faster than one call at a time.
const usageWorkerCount = 8

// usageMetricsByType maps each ARM resource type (lowercase) worth sampling
// to the Azure Monitor metric names that describe its actual utilization.
// Resource types not listed here are skipped — not every resource type
// exposes meaningful runtime metrics.
var usageMetricsByType = map[string][]string{
	"microsoft.documentdb/databaseaccounts":  {"TotalRequestUnits", "NormalizedRUConsumption"},
	"microsoft.storage/storageaccounts":      {"UsedCapacity", "Transactions"},
	"microsoft.web/serverfarms":              {"CpuPercentage", "MemoryPercentage"},
	"microsoft.web/sites":                    {"CpuTime", "Requests"},
	"microsoft.keyvault/vaults":              {"ServiceApiHit"},
	"microsoft.containerregistry/registries": {"TotalPullCount", "TotalPushCount"},
	"microsoft.network/publicipaddresses":    {"BytesInDDoS", "PacketsInDDoS"},
	"microsoft.cognitiveservices/accounts":   {"TotalCalls", "TotalErrors"},
}

// UsageData holds raw Azure Monitor metric values per resource. No waste
// scoring, savings estimates, or utilization judgments happen here — Claude
// interprets these numbers alongside the Cost extractor's data later.
type UsageData struct {
	TotalResourcesSampled int               `json:"total_resources_sampled"`
	PeriodFrom            string            `json:"period_from"`
	PeriodTo              string            `json:"period_to"`
	Metrics               []json.RawMessage `json:"metrics"`
}

// ExtractUsage discovers every resource of the supported types in the
// subscription and fetches Azure Monitor metrics for each. It is fully
// self-contained — it does not depend on the other 12 extractors' output,
// since CleanResource() strips the "id" field they'd otherwise need to
// reuse for a Monitor lookup.
func ExtractUsage(ctx context.Context, subID string, cred azcore.TokenCredential) (*UsageData, error) {
	resourcesClient, err := armresources.NewClient(subID, cred, nil)
	if err != nil {
		return nil, fmt.Errorf("creating resources client: %w", err)
	}
	metricsClient, err := armmonitor.NewMetricsClient(subID, cred, nil)
	if err != nil {
		return nil, fmt.Errorf("creating monitor client: %w", err)
	}

	end := time.Now().UTC()
	start := end.AddDate(0, 0, -30)
	timespan := fmt.Sprintf("%s/%s", start.Format(time.RFC3339), end.Format(time.RFC3339))

	// Build the full job list up front so all resource types can be sampled
	// concurrently, instead of processing one Monitor call at a time.
	type job struct {
		armType     string
		resourceID  string
		metricNames []string
	}
	var jobs []job

	for armType, metricNames := range usageMetricsByType {
		ids, err := listResourceIDsByType(ctx, resourcesClient, armType)
		if err != nil {
			fmt.Fprintf(os.Stderr, "  usage: listing %s failed: %v\n", armType, err)
			continue
		}
		fmt.Fprintf(os.Stderr, "  usage: %s — %d resource(s)\n", armType, len(ids))
		for _, id := range ids {
			jobs = append(jobs, job{armType, id, metricNames})
		}
	}

	jobCh := make(chan job)
	resultCh := make(chan []json.RawMessage)
	var wg sync.WaitGroup

	for w := 0; w < usageWorkerCount; w++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := range jobCh {
				metricCtx, cancel := context.WithTimeout(ctx, 20*time.Second)
				result, err := metricsClient.List(metricCtx, j.resourceID, &armmonitor.MetricsClientListOptions{
					Metricnames: strPtr(strings.Join(j.metricNames, ",")),
					Timespan:    strPtr(timespan),
					// Daily (not hourly) — the dashboard only ever displays a
					// period-level average/total, so 30 daily points is all
					// the resolution that's actually used. Hourly (720
					// points/metric) made usage_data over 1MB per audit and
					// took several seconds just to decompress on every page
					// load; daily cuts that ~24x while still leaving room for
					// a future day-by-day usage trend chart if one is built.
					Interval:    strPtr("P1D"),
					Aggregation: strPtr("Average,Total,Count"),
				})
				cancel()
				if err != nil {
					// One resource's metrics failing (e.g. metric not
					// supported for its SKU) shouldn't drop the others.
					resultCh <- nil
					continue
				}
				clean, err := cleanMetricResult(j.resourceID, result.Value)
				if err != nil {
					resultCh <- nil
					continue
				}
				resultCh <- clean
			}
		}()
	}

	go func() {
		for _, j := range jobs {
			jobCh <- j
		}
		close(jobCh)
	}()
	go func() {
		wg.Wait()
		close(resultCh)
	}()

	var entries []json.RawMessage
	for clean := range resultCh {
		entries = append(entries, clean...)
	}

	return &UsageData{
		TotalResourcesSampled: len(jobs),
		PeriodFrom:            start.Format("2006-01-02"),
		PeriodTo:              end.Format("2006-01-02"),
		Metrics:               entries,
	}, nil
}

// listResourceIDsByType lists every resource ARM ID of one resource type in
// the subscription.
func listResourceIDsByType(ctx context.Context, client *armresources.Client, armType string) ([]string, error) {
	filter := fmt.Sprintf("resourceType eq '%s'", armType)
	var ids []string

	pager := client.NewListPager(&armresources.ClientListOptions{Filter: &filter})
	for pager.More() {
		page, err := pager.NextPage(ctx)
		if err != nil {
			return ids, err
		}
		for _, r := range page.Value {
			if r.ID != nil {
				ids = append(ids, *r.ID)
			}
		}
	}
	return ids, nil
}

// cleanMetricResult reshapes Azure Monitor's per-metric response into one
// flat JSON object per metric, keeping every real datapoint (timestamp +
// average/total/count/min/max) exactly as returned. No thresholds, scores,
// or savings estimates are computed here.
func cleanMetricResult(resourceID string, metrics []*armmonitor.Metric) ([]json.RawMessage, error) {
	out := make([]json.RawMessage, 0, len(metrics))
	for _, m := range metrics {
		if m == nil {
			continue
		}

		name := ""
		if m.Name != nil && m.Name.Value != nil {
			name = *m.Name.Value
		}

		unit := ""
		if m.Unit != nil {
			unit = string(*m.Unit)
		}

		var dataPoints []map[string]any
		for _, ts := range m.Timeseries {
			if ts == nil {
				continue
			}
			for _, dp := range ts.Data {
				if dp == nil {
					continue
				}
				point := map[string]any{}
				if dp.TimeStamp != nil {
					point["timestamp"] = dp.TimeStamp.Format(time.RFC3339)
				}
				if dp.Average != nil {
					point["average"] = *dp.Average
				}
				if dp.Total != nil {
					point["total"] = *dp.Total
				}
				if dp.Count != nil {
					point["count"] = *dp.Count
				}
				if dp.Minimum != nil {
					point["minimum"] = *dp.Minimum
				}
				if dp.Maximum != nil {
					point["maximum"] = *dp.Maximum
				}
				dataPoints = append(dataPoints, point)
			}
		}

		// Pre-computed once here so every dashboard page load doesn't have to
		// re-loop over data_points just to get the same average/total.
		var avgSum float64
		var avgCount int
		var totalSum float64
		var totalCount int
		for _, p := range dataPoints {
			if v, ok := p["average"].(float64); ok {
				avgSum += v
				avgCount++
			}
			if v, ok := p["total"].(float64); ok {
				totalSum += v
				totalCount++
			}
		}
		summary := map[string]any{"avg": nil, "total": nil}
		if avgCount > 0 {
			summary["avg"] = avgSum / float64(avgCount)
		}
		if totalCount > 0 {
			summary["total"] = totalSum
		}

		entry := map[string]any{
			"resource_id": resourceID,
			"metric_name": name,
			"unit":        unit,
			"data_points": dataPoints,
			"summary":     summary,
		}
		clean, err := json.Marshal(entry)
		if err != nil {
			return nil, fmt.Errorf("marshaling metric entry: %w", err)
		}
		out = append(out, json.RawMessage(clean))
	}
	return out, nil
}

File: dashboard/app/api/utils/usage.ts
import { UsageMetricRaw } from '../types'

// Maps an ARM resource type (as embedded in the resource ID) to the same
// slug used elsewhere in the dashboard (icons, labels). Usage metrics only
// carry the full resource ID, not a clean type field, so the slug is
// recovered by parsing the ID's /providers/{Namespace}/{Type}/ segment.
const ARM_TYPE_TO_SLUG: Record<string, string> = {
  'microsoft.documentdb/databaseaccounts': 'cosmosdb',
  'microsoft.storage/storageaccounts': 'storage',
  'microsoft.web/serverfarms': 'appserviceplan',
  'microsoft.web/sites': 'appservice',
  'microsoft.keyvault/vaults': 'keyvault',
  'microsoft.containerregistry/registries': 'acr',
  'microsoft.network/publicipaddresses': 'publicip',
  'microsoft.cognitiveservices/accounts': 'cognitiveservices',
}

export function resourceTypeSlug(resourceId: string): string | null {
  const match = resourceId.match(/\/providers\/([^/]+)\/([^/]+)\//i)
  if (!match) return null
  return ARM_TYPE_TO_SLUG[`${match[1].toLowerCase()}/${match[2].toLowerCase()}`] || null
}

export function summarizeMetric(m: UsageMetricRaw) {
  // The CLI pre-computes this at collection time (see extractors/usage.go)
  // so reading it here is just a field lookup, not a re-scan of every
  // datapoint on every page load. Fall back to computing it for audits
  // collected before that field existed.
  let avg: number | null
  let total: number | null
  if (m.summary) {
    avg = m.summary.avg
    total = m.summary.total
  } else {
    const points = m.data_points || []
    const avgPoints = points.filter(p => p.average !== undefined)
    const totalPoints = points.filter(p => p.total !== undefined)
    avg = avgPoints.length ? avgPoints.reduce((s, p) => s + (p.average || 0), 0) / avgPoints.length : null
    total = totalPoints.length ? totalPoints.reduce((s, p) => s + (p.total || 0), 0) : null
  }
  return { metric_name: m.metric_name, unit: m.unit, avg, total, rank: total ?? avg ?? 0 }
}

export interface UsageGroup {
  resource_id: string
  metrics: { metric_name: string; unit: string; avg: number | null; total: number | null }[]
}

// Groups usage metrics for ONE resource type, sorted by activity — shared by
// the usage-summary endpoint (UI table) and the "usage:<type>" analyze/chat
// scope (LLM context), so both see identical numbers.
export function buildUsageGroups(metrics: UsageMetricRaw[], type: string): UsageGroup[] {
  const byResource = new Map<string, ReturnType<typeof summarizeMetric>[]>()
  for (const m of metrics) {
    if (resourceTypeSlug(m.resource_id) !== type) continue
    const list = byResource.get(m.resource_id) || []
    list.push(summarizeMetric(m))
    byResource.set(m.resource_id, list)
  }

  return Array.from(byResource.entries())
    .map(([resource_id, list]) => ({
      resource_id,
      rank: Math.max(0, ...list.map(m => m.rank)),
      metrics: list.map(m => ({ metric_name: m.metric_name, unit: m.unit, avg: m.avg, total: m.total })),
    }))
    .sort((a, b) => b.rank - a.rank)
    .map(({ resource_id, metrics }) => ({ resource_id, metrics }))
}

export interface UsageTypeInfo { slug: string; count: number }

// Lists which resource types have usage data and how many resources each —
// used to populate scope dropdowns (Analyze, Chat, Resource Utilization)
// without loading all usage data up front.
export function listUsageTypes(metrics: UsageMetricRaw[]): UsageTypeInfo[] {
  const resourcesByType = new Map<string, Set<string>>()
  for (const m of metrics) {
    const slug = resourceTypeSlug(m.resource_id)
    if (!slug) continue
    const set = resourcesByType.get(slug) || new Set<string>()
    set.add(m.resource_id)
    resourcesByType.set(slug, set)
  }
  return Array.from(resourcesByType.entries())
    .map(([slug, ids]) => ({ slug, count: ids.size }))
    .sort((a, b) => b.count - a.count)
}
